<?xml version="1.0"?>
<launch>

  <!-- "camera" should uniquely identify the device. All topics are pushed down
       into the "camera" namespace, and it is prepended to tf frame ids. -->
  <arg name="camera" default="camera"/>

  <arg name="frame_id" default="/$(arg camera)_optical_frame"/>

  <!-- Serial number of the device to open. If this argument is an empty string,
       the first available device will be opened. -->
  <arg name="device_serial" default=""/>

  <arg name="manager" value="$(arg camera)_nodelet_manager"/>

  <!-- Enable distance/depth/amplitude image rectification. If set to true,
       appropriate nodelets will be loaded. -->
  <arg name="rectify" default="true"/>

  <!-- Enable point cloud generation from the rectified depth image. If set to
       true, appropriate nodelet will be loaded.
       Note: the driver nodelet iself publishes point clouds which come from the
       PMD SDK driver in "points_unrectified" topic. As the name suggests, those
       point clouds are produced from distorted depth images. -->
  <arg name="points" default="true"/>

  <arg name="respawn" default="true"/>

  <!-- Start nodelet manager in top-level namespace -->
  <node pkg="nodelet" type="nodelet" name="$(arg manager)" output="screen"
        args="manager" />

  <!-- Push down all topics/nodelets into "camera" namespace -->
  <group ns="$(arg camera)">

    <!-- Driver nodelet -->
    <node pkg="nodelet" type="nodelet" name="driver" output="screen"
          args="load pmd_camboard_nano/driver /$(arg manager)">
          <param name="device_serial" value="$(arg device_serial)"/>
          <param name="frame_id" value="$(arg frame_id)"/>
    </node>

    <!-- Rectified distance image (optional) -->
    <node if="$(arg rectify)"
          pkg="nodelet" type="nodelet" name="rectify_distance"
          args="load image_proc/rectify /$(arg manager)"
          respawn="$(arg respawn)">
          <remap from="image_mono" to="distance/image"/>
          <remap from="camera_info" to="distance/camera_info"/>
          <remap from="image_rect" to="distance/image_rect"/>
          <!-- Use nearest neighbor (0) interpolation so we don't streak across
               depth boundaries -->
          <param name="interpolation" value="0"/>
    </node>

    <!-- Rectified depth image (optional) -->
    <node if="$(arg rectify)"
          pkg="nodelet" type="nodelet" name="rectify_depth"
          args="load image_proc/rectify /$(arg manager)"
          respawn="$(arg respawn)">
          <remap from="image_mono" to="depth/image"/>
          <remap from="camera_info" to="depth/camera_info"/>
          <remap from="image_rect" to="depth/image_rect"/>
          <!-- Use nearest neighbor (0) interpolation so we don't streak across
               depth boundaries -->
          <param name="interpolation" value="0"/>
    </node>

    <!-- Rectified amplitude image (optional) -->
    <node if="$(arg rectify)"
          pkg="nodelet" type="nodelet" name="rectify_amplitude"
          args="load image_proc/rectify /$(arg manager)"
          respawn="$(arg respawn)">
          <remap from="image_mono" to="amplitude/image"/>
          <remap from="camera_info" to="amplitude/camera_info"/>
          <remap from="image_rect" to="amplitude/image_rect"/>
    </node>

    <!-- XYZ point cloud (optional) -->
    <node if="$(arg points)"
          pkg="nodelet" type="nodelet" name="points"
          args="load depth_image_proc/point_cloud_xyz /$(arg manager)"
          respawn="$(arg respawn)">
          <remap from="image_rect" to="depth/image_rect"/>
          <remap from="camera_info" to="depth/camera_info"/>
          <remap from="points" to="points"/>
    </node>

  </group>

  <!-- Publish transforms for depth frame and depth optical frame -->
  <arg name="optical_rotate" value="0 0 0 -1.5707963267948966 0 -1.5707963267948966"/>

  <node pkg="tf" type="static_transform_publisher" name="$(arg camera)_base_link1"
        args="0 0 0 0 0 0 /$(arg camera)_link /$(arg camera)_frame 100"/>

  <node pkg="tf" type="static_transform_publisher" name="$(arg camera)_base_link2"
        args="$(arg optical_rotate) /$(arg camera)_frame /$(arg camera)_optical_frame 100"/>

</launch>
